### PR TITLE
create binary counter example

### LINKED ISSUE #48: Include simple example of binary counter
> I also think it would be very valuable to have a slightly extended example of how the counter works, where multiple elements are inserted. You could do an example where the elements are characters and the binary operator is min, for instance. 'zero' could be a space character. You could push in 8 letters in an order that causes multiple slots to be occupied at the end. You can use that to show that the counter only needs log2(8) = 3 slots. You can also show that after all the elements have been inserted, the job isn't yet finished. This justifies the need for the reduce_counter operation, which performs all the remaining comparisons to get to a single 'winner'. I think that's important, because with the text as it is currently, the reduction step is never really justified. It's just presented as something that is necessary.

### PR DESCRIPTION
**Pull request by justinmeiners**
fix #48 for @aharrison24

### CODE REVIEW

**Comment by aharrison24** [comment]
Looks good!

Making `binary_counter::counter` public for testing is understandable, but I wonder whether there's a cleaner way round the debugging requirement. You could consider providing `cbegin` and `cend` methods on `binary_counter` to allow callers to iterate over the elements directly. i.e. `counter.cbegin()` rather than the uglier `counter.counter.begin()`.

`show_counter` is essentially a `std::copy` into a `std::ostream_iterator<char>(std::cout, " ")`. You _could_ call `std::copy` directly from `main` and avoid needing to write the `show_counter` function. On the other hand, having the function raises the level of abstraction in `main`. And if you were going to replace the loop in the function with a `std::copy` then you'd need to start using `std::iterator_traits<I>::value_type` to deduce the underlying element type. On balance I suspect the raw for-loop in a function (as you have it) is cleaner!

**Comment by aharrison24** [comment]
Sorry, should have said that this is a really good example of how the binary counter should be used and what it does internally. Would have helped me a lot in my early understanding. Thank you!

**Comment by justinmeiners** [comment]
> public.

I agree it's not great. I was trying to avoid making any modifications, but I think mine is worse. I have added begin/end (I believe the cbegin/cend convention is C++11?).

> std::copy

I like this better, and chose to remove the function and make it inline.

File: code/test_binary_counter.cpp [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+template<typename T>
+struct min_op {
+  T operator()(const T& a, const T& b) {
+    return std::min(a, b);
+  }
+};
+
+int main() {
+  char letters[] = { 'C', 'H', 'B',  'F', 'I', 'D', 'E', 'G', 'A', 'J' };
```
Comment (aharrison24) [inline code/test_binary_counter.cpp:15]: I think an extra space char snuck in here.

Comment (justinmeiners) [reply code/test_binary_counter.cpp:15]: fixed

**Pull request merged**